src/plan_expand_hypertable.c

```diff
@@ -0,0 +1,253 @@
+#include <postgres.h>
+#include <nodes/relation.h>
+#include <parser/parsetree.h>
+#include <optimizer/var.h>
+#include <optimizer/restrictinfo.h>
+#include <catalog/pg_inherits_fn.h>
+#include <nodes/plannodes.h>
+#include <optimizer/prep.h>
+#include <nodes/nodeFuncs.h>
+
+#include "plan_expand_hypertable.h"
+#include "hypertable.h"
+#include "hypertable_restrict_info.h"
+#include "planner_import.h"
+#include "compat.h"
+
+
+
+typedef struct CollectQualCtx
+{
+	PlannerInfo *root;
+	RelOptInfo *rel;
+	List	   *result;
+} CollectQualCtx;
+
+
+static bool
+collect_quals_walker(Node *node, CollectQualCtx *ctx)
+{
+	if (node == NULL)
+		return false;
+
+	if (IsA(node, FromExpr))
+	{
+		FromExpr   *f = (FromExpr *) node;
+		ListCell   *lc;
+
+		foreach(lc, (List *) f->quals)
+		{
+			Node	   *qual = (Node *) lfirst(lc);
+			RestrictInfo *restrictinfo;
+
+			Relids		relids = pull_varnos(qual);
+
+			if (bms_num_members(relids) != 1 || !bms_is_member(ctx->rel->relid, relids))
+				continue;
+#if PG96
+			restrictinfo = make_restrictinfo((Expr *) qual,
+											 true,
+											 false,
+											 false,
+											 relids,
+											 NULL,
+											 NULL);
+#else
+			restrictinfo = make_restrictinfo((Expr *) qual,
+											 true,
+											 false,
+											 false,
+											 ctx->root->qual_security_level,
+											 relids,
+											 NULL,
+											 NULL);
+#endif
+			ctx->result = lappend(ctx->result, restrictinfo);
+		}
+	}
+
+	return expression_tree_walker(node, collect_quals_walker, ctx);
+}
+
+/* Since baserestrictinfo is not yet set by the planner, we have to derive
+ * it ourselves. It's safe for us to miss some restrict info clauses (this
+ * will just results in more chunks being included) so this does not need
+ * to be as comprehensive as the PG native derivation. This is inspired
+ * by the derivation in `deconstruct_recurse` in PG */
+
+static List *
+get_restrict_info(PlannerInfo *root, RelOptInfo *rel)
+{
+	CollectQualCtx ctx = {
+		.root = root,
+		.rel = rel,
+		.result = NIL,
+	};
+
+	collect_quals_walker((Node *) root->parse->jointree, &ctx);
+
+	return ctx.result;
+}
+
+static List *
+find_children_oids(HypertableRestrictInfo *hri, Hypertable *ht, LOCKMODE lockmode)
+{
+	List	   *result;
+
+	/*
+	 * optimization: using the HRI only makes sense if we are not using all
```

Comment:
Ambiguous comment: Is this optimization done now (doesn't look like it), or is it suggested?

Reply:
Fixed